File: scripts/chunk_markdown.py
#!/usr/bin/env python3
"""
Markdown chunker for docs/project-data.
Splits markdown files into semantic chunks based on heading structure.
Output: docs/project-data/chunks.jsonl
"""

import json
import re
import os
from pathlib import Path

BASE_DIR = Path("/Users/nikita.rybalov/workspace/proshop_mern/docs/project-data")
OUTPUT_FILE = BASE_DIR / "chunks.jsonl"

TARGET_TOKENS = 400
MAX_TOKENS = 600
MIN_TOKENS = 50


def estimate_tokens(text: str) -> int:
    """Rough token estimate: ~4 chars per token."""
    return len(text) // 4


def detect_language(text: str) -> str:
    """Detect language by checking for Cyrillic characters."""
    cyrillic = len(re.findall(r'[а-яА-ЯёЁ]', text))
    total_alpha = len(re.findall(r'[a-zA-Zа-яА-ЯёЁ]', text))
    if total_alpha == 0:
        return "en"
    return "ru" if cyrillic / total_alpha > 0.3 else "en"


def extract_keywords(text: str) -> list:
    """Extract keywords: function names, component names, tech terms."""
    keywords = set()

    # CamelCase / PascalCase words (component names)
    keywords.update(re.findall(r'\b([A-Z][a-z]+(?:[A-Z][a-z]+)+)\b', text))

    # camelCase (function/variable names)
    keywords.update(re.findall(r'\b([a-z]+(?:[A-Z][a-z]+)+)\b', text))

    # Backtick-wrapped code terms
    keywords.update(re.findall(r'`([^`]{2,30})`', text))

    # Known tech terms
    tech_terms = [
        'Redux', 'React', 'Express', 'MongoDB', 'Mongoose', 'JWT', 'PayPal',
        'Stripe', 'localStorage', 'axios', 'Node.js', 'REST', 'API',
        'middleware', 'controller', 'reducer', 'action', 'thunk', 'multer',
        'bcrypt', 'Bootstrap', 'Docker', 'nginx', 'PostgreSQL', 'pgvector'
    ]
    for term in tech_terms:
        if term.lower() in text.lower():
            keywords.add(term)

    # Limit to 7 most relevant
    return sorted(list(keywords))[:7]


def generate_summary(text: str, headings: list) -> str:
    """Generate a one-sentence summary from text content."""
    lines = [l.strip() for l in text.split('\n')
             if l.strip() and not l.strip().startswith('#')]
    if not lines:
        return " > ".join(headings) if headings else ""

    first_line = lines[0]
    # If first line is short enough, use it directly
    if len(first_line) <= 150:
        return first_line
    # Otherwise truncate at sentence boundary
    match = re.match(r'^(.{50,150}[.!?])\s', first_line)
    if match:
        return match.group(1)
    return first_line[:150] + "..."


def parse_markdown_sections(content: str) -> list:
    """Parse markdown into sections based on headings."""
    sections = []
    current_headings = {}  # level -> heading text
    current_text_lines = []
    current_level = 0

    lines = content.split('\n')

    for line in lines:
        heading_match = re.match(r'^(#{1,6})\s+(.+)$', line)
        if heading_match:
            # Save previous section
            if current_text_lines:
                text = '\n'.join(current_text_lines).strip()
                if text:
                    breadcrumbs = []
                    for lvl in sorted(current_headings.keys()):
                        if lvl > 1:
                            breadcrumbs.append(current_headings[lvl])
                    sections.append({
                        'text': text,
                        'headings': breadcrumbs,
                        'level': current_level
                    })
                current_text_lines = []

            level = len(heading_match.group(1))
            heading_text = heading_match.group(2).strip()
            current_level = level
            current_headings[level] = heading_text
            # Clear deeper headings
            for k in list(current_headings.keys()):
                if k > level:
                    del current_headings[k]
        else:
            current_text_lines.append(line)

    # Last section
    if current_text_lines:
        text = '\n'.join(current_text_lines).strip()
        if text:
            breadcrumbs = []
            for lvl in sorted(current_headings.keys()):
                if lvl > 1:
                    breadcrumbs.append(current_headings[lvl])
            sections.append({
                'text': text,
                'headings': breadcrumbs,
                'level': current_level
            })

    return sections


def split_large_section(text: str, target=TARGET_TOKENS) -> list:
    """Split a large section into chunks by paragraphs or table rows."""
    paragraphs = re.split(r'\n\n+', text)
    chunks = []
    current_chunk = []
    current_size = 0

    for para in paragraphs:
        para_tokens = estimate_tokens(para)

        # If a single paragraph (e.g. a table) is too large, split it
        if para_tokens > MAX_TOKENS:
            # Flush current chunk first
            if current_chunk:
                chunks.append('\n\n'.join(current_chunk))
                current_chunk = []
                current_size = 0
            # Check if it's a table
            para_lines = para.strip().split('\n')
            if any(re.match(r'^\|.*\|$', l.strip()) for l in para_lines[:3]):
                chunks.extend(split_table(para))
            else:
                # Split by lines into ~target-sized groups
                group = []
                group_size = 0
                for line in para_lines:
                    line_tokens = estimate_tokens(line)
                    if group_size + line_tokens > MAX_TOKENS and group:
                        chunks.append('\n'.join(group))
                        group = [line]
                        group_size = line_tokens
                    else:
                        group.append(line)
                        group_size += line_tokens
                if group:
                    chunks.append('\n'.join(group))
            continue

        if current_size + para_tokens > MAX_TOKENS and current_chunk:
            chunk_text = '\n\n'.join(current_chunk)
            chunks.append(chunk_text)
            # Overlap: keep last sentence of previous chunk
            last_sentences = re.findall(r'[^.!?]*[.!?]', current_chunk[-1])
            if last_sentences and len(last_sentences) > 1:
                overlap = last_sentences[-1].strip()
                current_chunk = [overlap, para]
                current_size = estimate_tokens(overlap) + para_tokens
            else:
                current_chunk = [para]
                current_size = para_tokens
        else:
            current_chunk.append(para)
            current_size += para_tokens

    if current_chunk:
        chunks.append('\n\n'.join(current_chunk))

    return chunks


def split_table(text: str) -> list:
    """Split a markdown table into chunks of rows, preserving header."""
    lines = text.split('\n')
    header_lines = []
    data_lines = []
    in_header = True

    for line in lines:
        if in_header:
            header_lines.append(line)
            if re.match(r'^\|[-:\s|]+\|$', line.strip()):
                in_header = False
        else:
            if line.strip():
                data_lines.append(line)

    if not data_lines:
        return [text]

    header = '\n'.join(header_lines)
    header_tokens = estimate_tokens(header)
    chunks = []
    current_rows = []
    current_size = header_tokens

    for row in data_lines:
        row_tokens = estimate_tokens(row)
        if current_size + row_tokens > MAX_TOKENS and current_rows:
            chunks.append(header + '\n' + '\n'.join(current_rows))
            current_rows = [row]
            current_size = header_tokens + row_tokens
        else:
            current_rows.append(row)
            current_size += row_tokens

    if current_rows:
        chunks.append(header + '\n' + '\n'.join(current_rows))

    return chunks


def get_title(content: str) -> str:
    """Extract title from first H1 heading."""
    match = re.match(r'^#\s+(.+)', content, re.MULTILINE)
    return match.group(1).strip() if match else ""


def process_file(filepath: Path) -> list:
    """Process a single markdown file into chunks."""
    content = filepath.read_text(encoding='utf-8')
    relative_path = str(filepath.relative_to(BASE_DIR.parent.parent))
    source_file = str(filepath.relative_to(BASE_DIR))
    title = get_title(content)

    sections = parse_markdown_sections(content)
    chunks = []

    # Merge tiny sections with next sibling
    merged_sections = []
    i = 0
    while i < len(sections):
        section = sections[i]
        tokens = estimate_tokens(section['text'])

        if tokens < MIN_TOKENS and i + 1 < len(sections):
            # Keep merging until we have enough
            merged_text = section['text']
            headings = section['headings']
            level = section['level']
            i += 1
            while i < len(sections) and estimate_tokens(merged_text) < MIN_TOKENS:
                merged_text += '\n\n' + sections[i]['text']
                if not headings:
                    headings = sections[i]['headings']
                i += 1
            merged_sections.append({
                'text': merged_text,
                'headings': headings,
                'level': level
            })
        else:
            merged_sections.append(section)
            i += 1

    for section in merged_sections:
        tokens = estimate_tokens(section['text'])

        if tokens <= MAX_TOKENS:
            chunks.append({
                'text': section['text'],
                'metadata': {
                    'source_file': source_file,
                    'file_path': relative_path,
                    'title': title,
                    'parent_headings': section['headings'],
                    'keywords': extract_keywords(section['text']),
                    'summary': generate_summary(section['text'], section['headings']),
                    'language': detect_language(section['text'])
                }
            })
        else:
            sub_chunks = split_large_section(section['text'])
            for chunk_text in sub_chunks:
                chunks.append({
                    'text': chunk_text,
                    'metadata': {
                        'source_file': source_file,
                        'file_path': relative_path,
                        'title': title,
                        'parent_headings': section['headings'],
                        'keywords': extract_keywords(chunk_text),
                        'summary': generate_summary(chunk_text, section['headings']),
                        'language': detect_language(chunk_text)
                    }
                })

    return chunks


def main():
    all_chunks = []
    md_files = sorted(BASE_DIR.rglob("*.md"))

    print(f"Processing {len(md_files)} markdown files...")

    for filepath in md_files:
        file_chunks = process_file(filepath)
        all_chunks.extend(file_chunks)
        print(f"  {filepath.relative_to(BASE_DIR)}: {len(file_chunks)} chunks")

    # Write JSONL
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        for chunk in all_chunks:
            f.write(json.dumps(chunk, ensure_ascii=False) + '\n')

    # Stats
    token_counts = [estimate_tokens(c['text']) for c in all_chunks]
    print(f"\n--- Results ---")
    print(f"Total chunks: {len(all_chunks)}")
    print(f"Output: {OUTPUT_FILE}")
    print(f"Token stats: min={min(token_counts)}, max={max(token_counts)}, "
          f"avg={sum(token_counts)//len(token_counts)}, "
          f"median={sorted(token_counts)[len(token_counts)//2]}")

    langs = {}
    for c in all_chunks:
        lang = c['metadata']['language']
        langs[lang] = langs.get(lang, 0) + 1
    print(f"Languages: {langs}")


if __name__ == "__main__":
    main()

File: scripts/chunk_markdown.js
#!/usr/bin/env node
/**
 * Markdown chunker for docs/project-data.
 * Splits markdown files into semantic chunks based on heading structure.
 * Output: docs/project-data/chunks.jsonl
 *
 * Usage: node scripts/chunk_markdown.js
 */

import { readFileSync, writeFileSync, readdirSync, statSync } from 'fs';
import { join, relative, resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const BASE_DIR = resolve(__dirname, '..', 'docs', 'project-data');
const OUTPUT_FILE = join(BASE_DIR, 'chunks.jsonl');

const TARGET_TOKENS = 400;
const MAX_TOKENS = 600;
const MIN_TOKENS = 50;

function estimateTokens(text) {
    return Math.floor(text.length / 4);
}

function detectLanguage(text) {
    const cyrillic = (text.match(/[а-яА-ЯёЁ]/g) || []).length;
    const totalAlpha = (text.match(/[a-zA-Zа-яА-ЯёЁ]/g) || []).length;
    if (totalAlpha === 0) return 'en';
    return cyrillic / totalAlpha > 0.3 ? 'ru' : 'en';
}

function extractKeywords(text) {
    const keywords = new Set();

    // PascalCase words (component names)
    const pascal = text.match(/\b([A-Z][a-z]+(?:[A-Z][a-z]+)+)\b/g);
    if (pascal) pascal.forEach((w) => keywords.add(w));

    // camelCase (function/variable names)
    const camel = text.match(/\b([a-z]+(?:[A-Z][a-z]+)+)\b/g);
    if (camel) camel.forEach((w) => keywords.add(w));

    // Backtick-wrapped code terms
    const backtick = text.match(/`([^`]{2,30})`/g);
    if (backtick) backtick.forEach((w) => keywords.add(w.replace(/`/g, '')));

    // Known tech terms
    const techTerms = [
        'Redux', 'React', 'Express', 'MongoDB', 'Mongoose', 'JWT', 'PayPal',
        'Stripe', 'localStorage', 'axios', 'Node.js', 'REST', 'API',
        'middleware', 'controller', 'reducer', 'action', 'thunk', 'multer',
        'bcrypt', 'Bootstrap', 'Docker', 'nginx', 'PostgreSQL', 'pgvector'
    ];
    const textLower = text.toLowerCase();
    for (const term of techTerms) {
        if (textLower.includes(term.toLowerCase())) {
            keywords.add(term);
        }
    }

    return [...keywords].sort().slice(0, 7);
}

function generateSummary(text, headings) {
    const lines = text.split('\n')
        .map((l) => l.trim())
        .filter((l) => l && !l.startsWith('#'));
    if (!lines.length) {
        return headings.length ? headings.join(' > ') : '';
    }

    const firstLine = lines[0];
    if (firstLine.length <= 150) return firstLine;

    const match = firstLine.match(/^(.{50,150}[.!?])\s/);
    if (match) return match[1];
    return firstLine.slice(0, 150) + '...';
}

function parseMarkdownSections(content) {
    const sections = [];
    const currentHeadings = {};
    let currentTextLines = [];
    let currentLevel = 0;

    const lines = content.split('\n');

    for (const line of lines) {
        const headingMatch = line.match(/^(#{1,6})\s+(.+)$/);
        if (headingMatch) {
            if (currentTextLines.length) {
                const text = currentTextLines.join('\n').trim();
                if (text) {
                    const breadcrumbs = Object.keys(currentHeadings)
                        .map(Number)
                        .sort((a, b) => a - b)
                        .filter((lvl) => lvl > 1)
                        .map((lvl) => currentHeadings[lvl]);
                    sections.push({ text, headings: breadcrumbs, level: currentLevel });
                }
                currentTextLines = [];
            }

            const level = headingMatch[1].length;
            const headingText = headingMatch[2].trim();
            currentLevel = level;
            currentHeadings[level] = headingText;
            // Clear deeper headings
            for (const k of Object.keys(currentHeadings).map(Number)) {
                if (k > level) delete currentHeadings[k];
            }
        } else {
            currentTextLines.push(line);
        }
    }

    // Last section
    if (currentTextLines.length) {
        const text = currentTextLines.join('\n').trim();
        if (text) {
            const breadcrumbs = Object.keys(currentHeadings)
                .map(Number)
                .sort((a, b) => a - b)
                .filter((lvl) => lvl > 1)
                .map((lvl) => currentHeadings[lvl]);
            sections.push({ text, headings: breadcrumbs, level: currentLevel });
        }
    }

    return sections;
}

function splitTable(text) {
    const lines = text.split('\n');
    const headerLines = [];
    const dataLines = [];
    let inHeader = true;

    for (const line of lines) {
        if (inHeader) {
            headerLines.push(line);
            if (/^\|[-:\s|]+\|$/.test(line.trim())) {
                inHeader = false;
            }
        } else if (line.trim()) {
            dataLines.push(line);
        }
    }

    if (!dataLines.length) return [text];

    const header = headerLines.join('\n');
    const headerTokens = estimateTokens(header);
    const chunks = [];
    let currentRows = [];
    let currentSize = headerTokens;

    for (const row of dataLines) {
        const rowTokens = estimateTokens(row);
        if (currentSize + rowTokens > MAX_TOKENS && currentRows.length) {
            chunks.push(header + '\n' + currentRows.join('\n'));
            currentRows = [row];
            currentSize = headerTokens + rowTokens;
        } else {
            currentRows.push(row);
            currentSize += rowTokens;
        }
    }

    if (currentRows.length) {
        chunks.push(header + '\n' + currentRows.join('\n'));
    }

    return chunks;
}

function splitLargeSection(text, target = TARGET_TOKENS) {
    const paragraphs = text.split(/\n\n+/);
    const chunks = [];
    let currentChunk = [];
    let currentSize = 0;

    for (const para of paragraphs) {
        const paraTokens = estimateTokens(para);

        // If a single paragraph is too large, split it
        if (paraTokens > MAX_TOKENS) {
            if (currentChunk.length) {
                chunks.push(currentChunk.join('\n\n'));
                currentChunk = [];
                currentSize = 0;
            }

            const paraLines = para.trim().split('\n');
            const isTable = paraLines.slice(0, 3).some((l) => /^\|.*\|$/.test(l.trim()));
            if (isTable) {
                chunks.push(...splitTable(para));
            } else {
                let group = [];
                let groupSize = 0;
                for (const line of paraLines) {
                    const lineTokens = estimateTokens(line);
                    if (groupSize + lineTokens > MAX_TOKENS && group.length) {
                        chunks.push(group.join('\n'));
                        group = [line];
                        groupSize = lineTokens;
                    } else {
                        group.push(line);
                        groupSize += lineTokens;
                    }
                }
                if (group.length) chunks.push(group.join('\n'));
            }
            continue;
        }

        if (currentSize + paraTokens > MAX_TOKENS && currentChunk.length) {
            const chunkText = currentChunk.join('\n\n');
            chunks.push(chunkText);
            // Overlap: keep last sentence of previous chunk
            const lastSentences = currentChunk[currentChunk.length - 1].match(/[^.!?]*[.!?]/g);
            if (lastSentences && lastSentences.length > 1) {
                const overlap = lastSentences[lastSentences.length - 1].trim();
                currentChunk = [overlap, para];
                currentSize = estimateTokens(overlap) + paraTokens;
            } else {
                currentChunk = [para];
                currentSize = paraTokens;
            }
        } else {
            currentChunk.push(para);
            currentSize += paraTokens;
        }
    }

    if (currentChunk.length) {
        chunks.push(currentChunk.join('\n\n'));
    }

    return chunks;
}

function getTitle(content) {
    const match = content.match(/^#\s+(.+)/m);
    return match ? match[1].trim() : '';
}

function findMarkdownFiles(dir) {
    const results = [];
    const entries = readdirSync(dir);
    for (const entry of entries) {
        const full = join(dir, entry);
        const stat = statSync(full);
        if (stat.isDirectory()) {
            results.push(...findMarkdownFiles(full));
        } else if (entry.endsWith('.md')) {
            results.push(full);
        }
    }
    return results.sort();
}

function processFile(filepath) {
    const content = readFileSync(filepath, 'utf-8');
    const relativePath = relative(resolve(__dirname, '..'), filepath);
    const sourceFile = relative(BASE_DIR, filepath);
    const title = getTitle(content);

    const sections = parseMarkdownSections(content);
    const chunks = [];

    // Merge tiny sections with next sibling
    const mergedSections = [];
    let i = 0;
    while (i < sections.length) {
        const section = sections[i];
        let tokens = estimateTokens(section.text);

        if (tokens < MIN_TOKENS && i + 1 < sections.length) {
            let mergedText = section.text;
            let headings = section.headings;
            const level = section.level;
            i++;
            while (i < sections.length && estimateTokens(mergedText) < MIN_TOKENS) {
                mergedText += '\n\n' + sections[i].text;
                if (!headings.length) headings = sections[i].headings;
                i++;
            }
            mergedSections.push({ text: mergedText, headings, level });
        } else {
            mergedSections.push(section);
            i++;
        }
    }

    for (const section of mergedSections) {
        const tokens = estimateTokens(section.text);
        const makeMeta = (text) => ({
            source_file: sourceFile,
            file_path: relativePath,
            title,
            parent_headings: section.headings,
            keywords: extractKeywords(text),
            summary: generateSummary(text, section.headings),
            language: detectLanguage(text)
        });

        if (tokens <= MAX_TOKENS) {
            chunks.push({ text: section.text, metadata: makeMeta(section.text) });
        } else {
            const subChunks = splitLargeSection(section.text);
            for (const chunkText of subChunks) {
                chunks.push({ text: chunkText, metadata: makeMeta(chunkText) });
            }
        }
    }

    return chunks;
}

function main() {
    const mdFiles = findMarkdownFiles(BASE_DIR);
    console.log(`Processing ${mdFiles.length} markdown files...`);

    const allChunks = [];
    for (const filepath of mdFiles) {
        const fileChunks = processFile(filepath);
        allChunks.push(...fileChunks);
        console.log(`  ${relative(BASE_DIR, filepath)}: ${fileChunks.length} chunks`);
    }

    // Write JSONL
    const lines = allChunks.map((c) => JSON.stringify(c));
    writeFileSync(OUTPUT_FILE, lines.join('\n') + '\n', 'utf-8');

    // Stats
    const tokenCounts = allChunks.map((c) => estimateTokens(c.text));
    const sorted = [...tokenCounts].sort((a, b) => a - b);
    console.log('\n--- Results ---');
    console.log(`Total chunks: ${allChunks.length}`);
    console.log(`Output: ${OUTPUT_FILE}`);
    console.log(`Token stats: min=${sorted[0]}, max=${sorted[sorted.length - 1]}, ` +
        `avg=${Math.floor(tokenCounts.reduce((a, b) => a + b, 0) / tokenCounts.length)}, ` +
        `median=${sorted[Math.floor(sorted.length / 2)]}`);

    const langs = {};
    for (const c of allChunks) {
        const lang = c.metadata.language;
        langs[lang] = (langs[lang] || 0) + 1;
    }
    console.log(`Languages: ${JSON.stringify(langs)}`);
}

main();
